File: Problems/94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque;
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = [];
        # Use an extra boolean to record if the node visited.
        stack = deque([(root, False)]);
        
        # Perform inorder traverse.
        while stack:
            cur, visited = stack.pop();
            # Terminal of tree nodes
            if not cur:
                continue;
            
            # Check if cur node seen.
            if visited:
                res.append(cur.val);
            else:
                # Add pending nodes
                stack.append((cur.right, False));
                stack.append((cur, True));
                stack.append((cur.left,False));
                
        return res;

File: Problems/94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.List;
import java.util.ArrayList;
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> record = new ArrayList(); // Record the result.
        if (root == null) return record;

        // Traverse left node.
        record.addAll(inorderTraversal(root.left));
        // Traverse mid node.
        record.add(root.val);
        // Traverse right node.
        record.addAll(inorderTraversal(root.right));
        
        return record;
    }
}